api/lightning/node_cln.py

```diff
@@ -0,0 +1,408 @@
+            }
+            return payout
+
+        # Some wallet providers (e.g. Muun) force routing through a private channel with high fees >1500ppm
+        # These payments will fail. So it is best to let the user know in advance this invoice is not valid.
+        route_hints = payreq_decoded.route_hints.hints
+
+        # Max amount RoboSats will pay for routing
+        if routing_budget_ppm == 0:
+            max_routing_fee_sats = max(
+                num_satoshis * float(config("PROPORTIONAL_ROUTING_FEE_LIMIT")),
+                float(config("MIN_FLAT_ROUTING_FEE_LIMIT_REWARD")),
+            )
+        else:
+            max_routing_fee_sats = int(
+                float(num_satoshis) * float(routing_budget_ppm) / 1000000
+            )
+
+        if route_hints:
+            routes_cost = []
+            # For every hinted route...
+            for hinted_route in route_hints:
+                route_cost = 0
+                # ...add up the cost of every hinted hop...
+                for hop_hint in hinted_route.hops:
+                    route_cost += hop_hint.feebase.msat / 1_000
+                    route_cost += (
+                        hop_hint.feeprop * num_satoshis / 1_000_000
+                    )
+
+                # ...and store the cost of the route to the array
+                routes_cost.append(route_cost)
+
+            # If the cheapest possible private route is more expensive than what RoboSats is willing to pay
+            if min(routes_cost) >= max_routing_fee_sats:
+                payout["context"] = {
+                    "bad_invoice": "The invoice hinted private routes are not payable within the submitted routing budget."
+                }
+                return payout
+
+        if payreq_decoded.amount_msat == 0:
+            payout["context"] = {
+                "bad_invoice": "The invoice provided has no explicit amount"
+            }
+            return payout
+
+        if not payreq_decoded.amount_msat // 1_000 == num_satoshis:
+            payout["context"] = {
+                "bad_invoice": "The invoice provided is not for "
+                + "{:,}".format(num_satoshis)
+                + " Sats"
+            }
+            return payout
+
+        payout["created_at"] = timezone.make_aware(
+            datetime.fromtimestamp(payreq_decoded.timestamp)
+        )
+        payout["expires_at"] = payout["created_at"] + timedelta(
+            seconds=payreq_decoded.expiry
+        )
+
+        if payout["expires_at"] < timezone.now():
+            payout["context"] = {
+                "bad_invoice": "The invoice provided has already expired"
+            }
+            return payout
+
+        payout["valid"] = True
+        payout["description"] = payreq_decoded.description
+        payout["payment_hash"] = payreq_decoded.payment_hash
+
+        return payout
+
+    @classmethod
+    def pay_invoice(cls, lnpayment):
```

Comment:
We have a pseudo-duplicate of this function outside the LNNode class. This one here used to pay out rewards. The other one, more sophisticated, is in `api/tasks` named `follow_send_payment(hash)`. 

While `api/lightning/node.py` pay_invoice() simply attempts to pay and expects a succeed/failed, the one in `tasks.py` evaluates every update during the routing (it uses an async celery worker thread). By using many workers and checking over all payments continuously the payout process is more robust to routing failures always keeping the user in the loop. There are also many shenanigans (e.g., payment is hodl by the recipient) that `follow_send_payment()` the celery worker task is safer for.

We will probably need to refactor and move the core gRPC functionality of `follow_send_payment()` into the `LNNode` class, so the celery task can use both CLN or LND.